### PR TITLE
Adding support for IoTeX blockchain #390

### PR DESCRIPTION
**Pull request by dustinxie**
## Description

This PR implemented #390 
One question: in src/Coin.cpp, I see TW::validateAddress() and TW::deriveAddress(), but don't see TW::curve(). It seems TW::curve() is only found in swift/wallet-core/Generated/Coins.cpp, but that is a generated file which is marked This is a GENERATED FILE, changes made here WILL BE LOST.
## Testing instructions
This PR includes unit tests for address and signing class/interface

## Checklist

- [ ] Prefix PR title with `[WIP]` if necessary.
- [ ] Add tests to cover changes as needed.
- [ ] Update documentation as needed.

### CODE REVIEW

File: include/TrustWalletCore/TWCoinType.h [hunk 1]
```diff
@@ -38,6 +38,7 @@ enum TWCoinType {
     TWCoinTypeICON = 74,
     TWCoinTypeIOST = 291,
     TWCoinTypeIocoin = 295,
+    TWCoinTypeIoTeX = 304,
```
Comment (dustinxie) [inline include/TrustWalletCore/TWCoinType.h]: IoTeX has index 304 and symbol IOTX, https://github.com/satoshilabs/slips/blob/master/slip-0044.md

File: src/IoTeX/Address.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    const auto dec = Bech32::decode(addr);
+    Bech32::convertBits<5, 8, false>(keyHash, dec.second);
+}
+
+Address::Address(const PublicKey& publicKey) {
+    if (publicKey.type != TWPublicKeyTypeSECP256k1Extended) {
+        throw std::invalid_argument("IoTeX: Address needs an extended SECP256k1 public key");
+    }
+    keyHash = publicKey.hash({}, static_cast<Data(*)(const byte*, const byte*)>(Hash::keccak256), true);
+    keyHash.erase(keyHash.begin(), keyHash.begin() + 12);
```
Comment (dustinxie) [inline src/IoTeX/Address.cpp]: public key hash is the last 20-byte of actual hash, similar to Ethereum. hence remove first 12 bytes

**Comment by alejandro-isaza** [comment]
Please add to coins.json. See the updated [Blockchain checklist](https://github.com/TrustWallet/wallet-core/blob/master/docs/Contributing.md#blockchain-checklist)

File: src/IoTeX/Address.cpp (hunk 2 above)
Comment (vikmeup) [reply src/IoTeX/Address.cpp]: is this bech32 address? then we have an address for it, `Bech32Address` or `TendermintAddress`

Comment (dustinxie) [reply src/IoTeX/Address.cpp]: yes, the address is Bech32 encoding of the public key hash, you can please refer to Address::string()

Comment (dustinxie) [reply src/IoTeX/Address.cpp]: will update the PR soon, thx

Comment (vikmeup) [reply src/IoTeX/Address.cpp]: Just updated, it's called `CosmosAddress`, and there is also `SegwitAddress` which is bech32 for bitcoin

**Comment by dustinxie** [comment]
> Please add to coins.json. See the updated [Blockchain checklist](https://github.com/TrustWallet/wallet-core/blob/master/docs/NewBlockchain.md)

done in new commit 936147e

File: src/IoTeX/Address.cpp (hunk 2 above)
Comment (dustinxie) [reply src/IoTeX/Address.cpp]: yes you are right, our address is similar, with different hrp
the implementation referenced these existing code

**Comment by alejandro-isaza** [comment]
Build is failing for iOS and Android, please fix.

**Comment by dustinxie** [comment]
> Build is failing for iOS and Android, please fix.

Android build is due to a test, I know how to fix. Could you shed light on why my iOS build is failing? I changed CoinType+Address.swift by adding
> case .iotex:
>            return IoTeXAddress(string: string)
this is follow suite today's EOSIO check-in, but build log complains:
> /Users/vsts/agent/2.150.3/work/1/s/swift/Sources/Addresses/CoinType+Address.swift:79:15: error: pattern cannot match values of type 'CoinType'
>        case .iotex:
I think something is missing but compared with today's EOSIO check-in, cannot seem to find what I am missing

**Comment by alejandro-isaza** [comment]
should be `.ioTeX`

**Comment by dustinxie** [comment]
> should be `.ioTeX`

thx, i just pushed the PR again. Just curious, what is the rule here? it is == "name" in coins.json with first letter lower-case? that seems to be the case for IoTeX, VeChain, BravoCoin, etc.

but for EOS, NEO, it is == "id" of coins.json, the code has ".eos", ".neo" but not .eOS or .nEO, kind of confused on this

**Comment by alejandro-isaza** [comment]
Swift cases are lowercase. So all caps becomes all lowercase. And mixed case becomes lowercase just for the first part. See https://swift.org/documentation/api-design-guidelines/

**Comment by dustinxie** [comment]
> Swift cases are lowercase. So all caps becomes all lowercase. And mixed case becomes lowercase just for the first part. See https://swift.org/documentation/api-design-guidelines/. Specifically "Acronyms and initialisms that commonly appear as all upper case in American English should be uniformly up- or down-cased according to case conventions:"
> 
> > Swift.org**[Swift.org](https://swift.org)**Swift is a general-purpose programming language built using a modern approach to safety, performance, and software design patterns.

thanks, good to learn

**Comment by dustinxie** [comment]
Android build was passing in last push and timed out in this build. 
iOS build is fixed in this push.
Can you please merge this PR?

**Pull request merged**